Source/Evolve/WarpXEvolveImplicitPicard.cpp

```diff
@@ -0,0 +1,580 @@
+/* Copyright 2022 David Grote
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+#include "WarpX.H"
+
+#include "BoundaryConditions/PML.H"
+#include "Diagnostics/MultiDiagnostics.H"
+#include "Diagnostics/ReducedDiags/MultiReducedDiags.H"
+#include "Evolve/WarpXDtType.H"
+#include "Evolve/WarpXPushType.H"
+#ifdef WARPX_USE_PSATD
+#   ifdef WARPX_DIM_RZ
+#       include "FieldSolver/SpectralSolver/SpectralSolverRZ.H"
+#   else
+#       include "FieldSolver/SpectralSolver/SpectralSolver.H"
+#   endif
+#endif
+#include "Parallelization/GuardCellManager.H"
+#include "Particles/MultiParticleContainer.H"
+#include "Particles/ParticleBoundaryBuffer.H"
+#include "Python/callbacks.H"
+#include "Utils/TextMsg.H"
+#include "Utils/WarpXAlgorithmSelection.H"
+#include "Utils/WarpXUtil.H"
+#include "Utils/WarpXConst.H"
+#include "Utils/WarpXProfilerWrapper.H"
+
+#include <ablastr/utils/SignalHandling.H>
+#include <ablastr/warn_manager/WarnManager.H>
+
+#include <AMReX.H>
+#include <AMReX_Array.H>
+#include <AMReX_BLassert.H>
+#include <AMReX_Geometry.H>
+#include <AMReX_IntVect.H>
+#include <AMReX_LayoutData.H>
+#include <AMReX_MultiFab.H>
+#include <AMReX_ParmParse.H>
+#include <AMReX_Print.H>
+#include <AMReX_REAL.H>
+#include <AMReX_Utility.H>
+#include <AMReX_Vector.H>
+
+#include <algorithm>
+#include <array>
+#include <memory>
+#include <ostream>
+#include <vector>
+
+using namespace amrex;
+using ablastr::utils::SignalHandling;
+
+void
+WarpX::EvolveImplicitPicardInit (int lev)
+{
+
+    if (lev == 0) {
+        // Add space to save the positions and velocities at the start of the time steps
+        for (auto const& pc : *mypc) {
+#if (AMREX_SPACEDIM >= 2)
+            pc->AddRealComp("x_n");
+#endif
+#if defined(WARPX_DIM_3D) || defined(WARPX_DIM_RZ)
+            pc->AddRealComp("y_n");
+#endif
+            pc->AddRealComp("z_n");
+            pc->AddRealComp("ux_n");
+            pc->AddRealComp("uy_n");
+            pc->AddRealComp("uz_n");
+        }
+    }
+
+    // Initialize MultiFabs to hold the E and B fields at the start of the time steps
+    // Only one refinement level is supported
+    const int nlevs_max = maxLevel() + 1;
+    Efield_n.resize(nlevs_max);
+    Efield_save.resize(nlevs_max);
+    if (evolve_scheme == EvolveScheme::ImplicitPicard) {
+        Bfield_n.resize(nlevs_max);
+        Bfield_save.resize(nlevs_max);
+    }
+
+    // Strange, the WarpX::DistributionMap(0) is not consistent with Ex_fp.DistributionMap()???
```

Comment:
Is this comment still up to date?

Reply:
The comment would still be true but is not needed here. I deleted it.